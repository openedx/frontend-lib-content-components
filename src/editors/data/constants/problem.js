import { StrictDict } from '../../utils';

export const ProblemTypeKeys = StrictDict({
  TEXTINPUT: 'TEXTINPUT',
  NUMERIC: 'NUMERIC',
  DROPDOWN: 'DROPDOWN',
  MULTISELECT: 'MULTISELECT',
  SINGLESELECT: 'SINGLESELECT',
});

export const ProblemTypes = StrictDict({
  [ProblemTypeKeys.SINGLESELECT]: {
    title: 'Single Select Problem',
    preview: ('<div />'),
    description: 'Specify one correct answer from a list of possible options',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.MULTISELECT]: {
    title: 'Multi Select Problem',
    preview: ('<div />'),
    description: 'Specify one or more correct answers from a list of possible options.',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.DROPDOWN]: {
    title: 'Dropdown Problem',
    preview: ('<div />'),
    description: 'Specify one correct answer from a list of possible options, selected in a dropdown menu.',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.NUMERIC]: {
    title: 'Numeric Response Problem',
    preview: ('<div />'),
    description: 'Specify one or more correct numeric answers, submitted in a response field.',
    helpLink: 'something.com',
  },
  [ProblemTypeKeys.TEXTINPUT]: {
    title: 'Text Input Problem',
    preview: ('<div />'),
    description: 'Specify one or more correct text answers, including numbers and special characters, submitted in a response field.',
    helpLink: 'something.com',
  },
});

export const ShowAnswerTypes = StrictDict({
  ALWAYS: {
      name: 'Always',
      value: 'always',
  },
  ANSWERED: {
      name: 'Answered',
      value: 'answered',
  },
  ATTEMPTED: {
      name: 'Attempted or Past Due',
      value: 'attempted',
  },
  CLOSED: {
      name: 'Closed',
      value: 'closed',
  },
  FINISHED: {
      name: 'Finished',
      value: 'finished',
  },
  CORRECT_OR_PAST_DUE: {
      name: 'Correct or Past Due',
      value: 'correct_or_past_due',
  },
  PAST_DUE: {
      name: 'Past Due',
      value: 'past_due',
  },
  NEVER: {
      name: 'Never',
      value: 'never',
  },
  AFTER_SOME_NUMBER_OF_ATTEMPTS: {
      name: 'After Some Number of Attempts',
      value: 'after_attempts',
  },
  AFTER_ALL_ATTEMPTS: {
      name: 'After All Attempts',
      value: 'after_all_attempts',
  },
  AFTER_ALL_ATTEMPTS_OR_CORRECT: {
      name: 'After All Attempts or Correct',
      value: 'after_all_attempts_or_correct',
  },
  ATTEMPTED_NO_PAST_DUE: {
      name: 'Attempted',
      value: 'attempted_no_past_due',
  },
});


export const RandomizationType = StrictDict({
  ALWAYS: {
      name: 'Always',
      value: 'always',
  },
  ONRESET: {
      name: 'On Reset',
      value: 'onreset',
  },
  NEVER: {
      name: 'Never',
      value: 'never',
  },
  PER_STUDENT: {
      name: 'Per Student',
      value: 'per_student',
  },
});

  
