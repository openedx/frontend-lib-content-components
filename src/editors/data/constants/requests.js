import { StrictDict } from '../../utils';

export const RequestStates = StrictDict({
  inactive: 'inactive',
  pending: 'pending',
  completed: 'completed',
  failed: 'failed',
});

export const RequestKeys = StrictDict({
  fetchImages: 'fetchImages',
  fetchVideos: 'fetchVideos',
  fetchBlock: 'fetchBlock',
  fetchImages: 'fetchImages',
  fetchUnit: 'fetchUnit',
  fetchStudioView: 'fetchStudioView',
  saveBlock: 'saveBlock',
  uploadAsset: 'uploadAsset',
  uploadVideo: 'uploadVideo',
  allowThumbnailUpload: 'allowThumbnailUpload',
  uploadThumbnail: 'uploadThumbnail',
  uploadTranscript: 'uploadTranscript',
  deleteTranscript: 'deleteTranscript',
  fetchCourseDetails: 'fetchCourseDetails',
  updateTranscriptLanguage: 'updateTranscriptLanguage',
  getTranscriptFile: 'getTranscriptFile',
  checkTranscriptsForImport: 'checkTranscriptsForImport',
  importTranscript: 'importTranscript',
  uploadAsset: 'uploadAsset',
  fetchAdvancedSettings: 'fetchAdvancedSettings',
  fetchVideoFeatures: 'fetchVideoFeatures',
});
