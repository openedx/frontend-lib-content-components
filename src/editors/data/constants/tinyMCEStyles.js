const getStyles = () => (
  `@import url("https://fonts.googleapis.com/css2?family=Open+Sans:ital,wght@0,300;0,400;0,600;0,700;1,300;1,400;1,600;1,700&display=swap");

  .mce-content-body *[contentEditable=false] {
    cursor: default;
  }
  .mce-content-body *[contentEditable=true] {
    cursor: text;
  }
  .mce-content-body div.mce-resizehandle {
    background-color: #4099ff;
    border-color: #4099ff;
    border-style: solid;
    border-width: 1px;
    box-sizing: border-box;
    height: 10px;
    position: absolute;
    width: 10px;
    z-index: 1298;
  }
  .mce-content-body div.mce-resizehandle:hover {
    background-color: #4099ff;
  }
  .mce-content-body div.mce-resizehandle:nth-of-type(1) {
    cursor: nwse-resize;
  }
  .mce-content-body div.mce-resizehandle:nth-of-type(2) {
    cursor: nesw-resize;
  }
  .mce-content-body div.mce-resizehandle:nth-of-type(3) {
    cursor: nwse-resize;
  }
  .mce-content-body div.mce-resizehandle:nth-of-type(4) {
    cursor: nesw-resize;
  }
  .mce-content-body .mce-resize-backdrop {
    z-index: 10000;
  }
  .mce-content-body .mce-clonedresizable {
    cursor: default;
    opacity: 0.5;
    outline: 1px dashed black;
    position: absolute;
    z-index: 10001;
  }
  .mce-content-body .mce-clonedresizable.mce-resizetable-columns th,
  .mce-content-body .mce-clonedresizable.mce-resizetable-columns td {
    border: 0;
  }
  .mce-content-body .mce-resize-helper {
    background: #555;
    background: rgba(0, 0, 0, 0.75);
    border: 1px;
    border-radius: 3px;
    color: white;
    display: none;
    font-family: sans-serif;
    font-size: 12px;
    line-height: 14px;
    margin: 5px 10px;
    padding: 5px;
    position: absolute;
    white-space: nowrap;
    z-index: 10002;
  }
  .mce-content-body img[data-mce-selected],
  .mce-content-body video[data-mce-selected],
  .mce-content-body audio[data-mce-selected],
  .mce-content-body object[data-mce-selected],
  .mce-content-body embed[data-mce-selected],
  .mce-content-body table[data-mce-selected] {
    outline: 3px solid #b4d7ff;
  }
  .mce-content-body *[contentEditable=false] *[contentEditable=true]:focus {
    outline: 3px solid #b4d7ff;
  }
  .mce-content-body *[contentEditable=false] *[contentEditable=true]:hover {
    outline: 3px solid #b4d7ff;
  }
  .mce-content-body *[contentEditable=false][data-mce-selected] {
    cursor: not-allowed;
    outline: 3px solid #b4d7ff;
  }
  .mce-content-body.mce-content-readonly *[contentEditable=true]:focus,
  .mce-content-body.mce-content-readonly *[contentEditable=true]:hover {
    outline: none;
  }
  .mce-content-body *[data-mce-selected="inline-boundary"] {
    background-color: #b4d7ff;
  }
  .mce-content-body .mce-edit-focus {
    outline: 3px solid #b4d7ff;
  }
  .mce-content-body img::-moz-selection {
    background: none;
  }
  .mce-content-body img::selection {
    background: none;
  }
  .mce-content-body {
      padding: 10px;
      background-color: #fff;
      font-family: 'Open Sans', Verdana, Arial, Helvetica, sans-serif;
      font-size: 16px;
      line-height: 1.6;
      color: #3c3c3c;
      scrollbar-3dlight-color: #F0F0EE;
      scrollbar-arrow-color: #676662;
      scrollbar-base-color: #F0F0EE;
      scrollbar-darkshadow-color: #DDDDDD;
      scrollbar-face-color: #E0E0DD;
      scrollbar-highlight-color: #F0F0EE;
      scrollbar-shadow-color: #F0F0EE;
      scrollbar-track-color: #F5F5F5;
  }
  .mce-content-body h1,
  .mce-content-body .hd-1 {
      color: #3c3c3c;
      font-weight: normal;
      font-size: 2em;
      line-height: 1.4em;
      margin: 0 0 1.41575em 0;
  }
  .mce-content-body h2,
  .mce-content-body .hd-2 {
      letter-spacing: 1px;
      margin-bottom: 15px;
      color: #646464;
      font-weight: 300;
      font-size: 1.2em;
      line-height: 1.2em;
      text-transform: uppercase;
  }
  .mce-content-body h3,
  .mce-content-body .hd-3 {
    margin: 0 0 10px 0;
    font-size: 1.1125em;
    font-weight: 400;
    text-transform: initial;
  }
  .mce-content-body .hd-3,
  .mce-content-body h4,
  .mce-content-body .hd-4,
  .mce-content-body h5,
  .mce-content-body .hd-5,
  .mce-content-body h6,
  .mce-content-body .hd-6 {
      margin: 0 0 10px 0;
      font-weight: 600;
  }
  .mce-content-body h4,
  .mce-content-body .hd-4 {
    font-size: 1em;
  }
  .mce-content-body h5,
  .mce-content-body .hd-5 {
      font-size: 0.83em;
  }
  .mce-content-body h6,
  .mce-content-body .hd-6 {
      font-size: 0.75em;
  }
  .mce-content-body p {
      margin-bottom: 1.416em;
      font-size: 1em;
      line-height: 1.6em !important;
      color: #3c3c3c;
  }
  .mce-content-body em, .mce-content-body i {
      font-style: italic;
  }
  .mce-content-body strong, .mce-content-body b {
      font-weight: bold;
  }
  .mce-content-body p + p, .mce-content-body ul + p, .mce-content-body ol + p {
      margin-top: 20px;
  }
  .mce-content-body ol, .mce-content-body ul {
      margin: 1em 0;
      padding: 0 0 0 1em;
      color: #3c3c3c;
  }
  .mce-content-body ol li, .mce-content-body ul li {
      margin-bottom: 0.708em;
  }
  .mce-content-body ol {
      list-style: decimal outside none;
      margin: 0;
  }
  .mce-content-body ul {
      list-style: disc outside none;
      margin: 0;
  }
  .mce-content-body a, .mce-content-body a:link, .mce-content-body a:visited, .mce-content-body a:hover, .mce-content-body a:active {
      color: #0075b4;
      text-decoration: none;
  }
  .mce-content-body img {
      max-width: 100%;
      height: auto;
  }
  .mce-content-body pre {
    margin: 1em 0;
    color: #3c3c3c;
    font-family: monospace, serif;
    font-size: 1em;
    white-space: pre-wrap;
    word-wrap: break-word;
  }
  .mce-content-body code {
      font-family: monospace, serif;
      background: none;
      color: #3c3c3c;
      padding: 0;
  }`
);

export { getStyles };

export default getStyles({});
