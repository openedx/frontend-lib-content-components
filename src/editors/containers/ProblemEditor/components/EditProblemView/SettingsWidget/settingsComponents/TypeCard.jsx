import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import SettingsOption from '../SettingsOption';
import { ProblemTypeKeys, ProblemTypes } from '../../../../../../data/constants/problem';
import messages from '../messages';
import TypeRow from './TypeRow';

export const TypeCard = ({
  answers,
  correctAnswerCount,
  problemType,
  updateField,
  updateAnswer,
  // inject
  intl,
}) => {
  const problemTypeKeysArray = Object.values(ProblemTypeKeys).filter(key => key !== ProblemTypeKeys.ADVANCED);

  if (problemType === ProblemTypeKeys.ADVANCED) { return null; }

  return (
    <SettingsOption
      title={intl.formatMessage(messages.typeSettingTitle)}
      summary={ProblemTypes[problemType].title}
    >
      {problemTypeKeysArray.map((typeKey, i) => (
        <TypeRow
          answers={answers}
          correctAnswerCount={correctAnswerCount}
          key={typeKey}
          typeKey={typeKey}
          label={ProblemTypes[typeKey].title}
          selected={typeKey !== problemType}
          lastRow={(i + 1) === problemTypeKeysArray.length}
          updateField={updateField}
          updateAnswer={updateAnswer}
        />
      ))}
    </SettingsOption>
  );
};

TypeCard.propTypes = {
  answers: PropTypes.arrayOf(PropTypes.shape({
    correct: PropTypes.bool,
    id: PropTypes.string,
    selectedFeedback: PropTypes.string,
    title: PropTypes.string,
    unselectedFeedback: PropTypes.string,
  })).isRequired,
  correctAnswerCount: PropTypes.number.isRequired,
  problemType: PropTypes.string.isRequired,
  updateField: PropTypes.func.isRequired,
  updateAnswer: PropTypes.func.isRequired,
  // injected
  intl: intlShape.isRequired,
};

export default injectIntl(TypeCard);
