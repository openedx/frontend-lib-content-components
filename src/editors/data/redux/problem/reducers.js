import _ from 'lodash-es';
import { createSlice } from '@reduxjs/toolkit';
import { indexToLetterMap } from '../../../containers/ProblemEditor/data/OLXParser';
import { StrictDict } from '../../../utils';
import { ProblemTypeKeys, ShowAnswerTypesKeys } from '../../constants/problem';

const nextAlphaId = (lastId) => String.fromCharCode(lastId.charCodeAt(0) + 1);
const initialState = {
  rawOLX: '',
  problemType: null,
  question: '',
  answers: [],
  correctAnswerCount: 0,
  groupFeedbackList: [],
  additionalAttributes: {},
  settings: {
    scoring: {
      weight: 0,
      attempts: {
        unlimited: true,
        number: null,
      },
    },
    hints: [],
    timeBetween: 0,
    matLabApiKey: '',
    showAnswer: {
      on: ShowAnswerTypesKeys.FINISHED,
      afterAttempts: 0,
    },
    showResetButton: false,
  },
};

// eslint-disable-next-line no-unused-vars
const problem = createSlice({
  name: 'problem',
  initialState,
  reducers: {
    updateField: (state, { payload }) => ({
      ...state,
      ...payload,
    }),
    updateQuestion: (state, { payload }) => ({
      ...state,
      question: payload,
    }),
    updateAnswer: (state, { payload }) => {
      const { id, hasSingleAnswer, ...answer } = payload;
      let { correctAnswerCount } = state;
      const answers = state.answers.map(obj => {
        if (obj.id === id) {
          if (_.has(answer, 'correct') && payload.correct) {
            correctAnswerCount += 1;
            return { ...obj, ...answer };
          }
          if (_.has(answer, 'correct') && payload.correct === false) {
            correctAnswerCount -= 1;
            return { ...obj, ...answer };
          }
          return { ...obj, ...answer };
        }
        // set other answers as incorrect if problem only has one answer correct
        // and changes object include correct key change
        if (hasSingleAnswer && _.has(answer, 'correct') && obj.correct) {
          return { ...obj, correct: false };
        }
        return obj;
      });
      return {
        ...state,
        correctAnswerCount,
        answers,
      };
    },
    deleteAnswer: (state, { payload }) => {
      const { id, correct } = payload;
      if (state.answers.length <= 1) {
        return state;
      }
      let { correctAnswerCount } = state;
      if (correct) {
        correctAnswerCount -= 1;
      }
      const answers = state.answers.filter(obj => obj.id !== id).map((answer, index) => {
        const newId = indexToLetterMap[index];
        if (answer.id === newId) {
          return answer;
        }
        return { ...answer, id: newId };
      });
      return {
        ...state,
        correctAnswerCount,
        answers,
      };
    },
    addAnswer: (state) => {
      const currAnswers = state.answers;
      if (currAnswers.length >= indexToLetterMap.length) {
        return state;
      }
      const newOption = {
        id: currAnswers.length ? nextAlphaId(currAnswers[currAnswers.length - 1].id) : 'A',
        title: '',
        selectedFeedback: undefined,
        unselectedFeedback: undefined,
        correct: false,
      };
      if (state.problemType === ProblemTypeKeys.MULTISELECT) {
        newOption.unselectedFeedback = '';
      }
      newOption.selectedFeedback = '';
      const answers = [
        ...currAnswers,
        newOption,
      ];
      return {
        ...state,
        answers,
      };
    },
    updateSettings: (state, { payload }) => ({
      ...state,
      settings: {
        ...state.settings,
        ...payload,
      },
    }),
    load: (state, { payload: { settings: { scoring, showAnswer, ...settings }, ...payload } }) => ({
      ...state,
      settings: {
        ...state.settings,
        scoring: { ...state.settings.scoring, ...scoring },
        showAnswer: { ...state.settings.showAnswer, ...showAnswer },
        ...settings,
      },
      ...payload,
    }),
    setEnableTypeSelection: (state) => ({
      ...state,
      problemType: null,
    }),
  },
});

const actions = StrictDict(problem.actions);

const { reducer } = problem;

export {
  actions,
  initialState,
  reducer,
};
