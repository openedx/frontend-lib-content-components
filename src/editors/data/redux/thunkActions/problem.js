import _ from 'lodash-es';
/* eslint-disable import/no-cycle */
import { actions } from '..';
import * as requests from './requests';
import { OLXParser } from '../../../containers/ProblemEditor/data/OLXParser';
import { parseSettings } from '../../../containers/ProblemEditor/data/SettingsParser';
import { ProblemTypeKeys } from '../../constants/problem';
import ReactStateOLXParser from '../../../containers/ProblemEditor/data/ReactStateOLXParser';
import { blankProblemOLX } from '../../../containers/ProblemEditor/data/mockData/olxTestData';
import { camelizeKeys } from '../../../utils';
import { fetchEditorContent } from '../../../containers/ProblemEditor/components/EditProblemView/hooks';

export const switchToAdvancedEditor = () => (dispatch, getState) => {
  const state = getState();
  const editorObject = fetchEditorContent({ format: '' });
  const reactOLXParser = new ReactStateOLXParser({ problem: state.problem, editorObject });
  const rawOLX = reactOLXParser.buildOLX();
  dispatch(actions.problem.updateField({ problemType: ProblemTypeKeys.ADVANCED, rawOLX }));
};

export const isBlankProblem = ({ rawOLX }) => {
  if (rawOLX.replace(/\s/g, '') === blankProblemOLX.rawOLX) {
    return true;
  }
  return false;
};

export const getDataFromOlx = ({ rawOLX, rawSettings, defaultSettings }) => {
  let olxParser;
  let parsedProblem;
  try {
    olxParser = new OLXParser(rawOLX);
    parsedProblem = olxParser.getParsedOLXData();
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error('The Problem Could Not Be Parsed from OLX. redirecting to Advanced editor.', error);
    return { problemType: ProblemTypeKeys.ADVANCED, rawOLX, settings: parseSettings(rawSettings, defaultSettings) };
  }
  if (parsedProblem?.problemType === ProblemTypeKeys.ADVANCED) {
    return { problemType: ProblemTypeKeys.ADVANCED, rawOLX, settings: parseSettings(rawSettings, defaultSettings) };
  }
  const { settings, ...data } = parsedProblem;
  const parsedSettings = { ...settings, ...parseSettings(rawSettings, defaultSettings) };
  if (!_.isEmpty(rawOLX) && !_.isEmpty(data)) {
    return { ...data, rawOLX, settings: parsedSettings };
  }
  return { settings: parsedSettings };
};

export const loadProblem = ({ rawOLX, rawSettings, defaultSettings }) => (dispatch) => {
  if (isBlankProblem({ rawOLX })) {
    dispatch(actions.problem.setEnableTypeSelection(camelizeKeys(defaultSettings)));
  } else {
    dispatch(actions.problem.load(getDataFromOlx({ rawOLX, rawSettings, defaultSettings })));
  }
};

export const fetchAdvancedSettings = ({ rawOLX, rawSettings }) => (dispatch) => {
  const advancedProblemSettingKeys = ['max_attempts', 'showanswer', 'show_reset_button', 'rerandomize'];
  dispatch(requests.fetchAdvancedSettings({
    onSuccess: (response) => {
      const defaultSettings = {};
      Object.entries(response.data).forEach(([key, value]) => {
        if (advancedProblemSettingKeys.includes(key)) {
          defaultSettings[key] = value.value;
        }
      });
      dispatch(actions.problem.updateField({ defaultSettings: camelizeKeys(defaultSettings) }));
      loadProblem({ rawOLX, rawSettings, defaultSettings })(dispatch);
    },
    onFailure: () => { loadProblem({ rawOLX, rawSettings, defaultSettings: {} })(dispatch); },
  }));
};

export const initializeProblem = (blockValue) => (dispatch) => {
  const rawOLX = _.get(blockValue, 'data.data', {});
  const rawSettings = _.get(blockValue, 'data.metadata', {});
  dispatch(fetchAdvancedSettings({ rawOLX, rawSettings }));
};

export default { initializeProblem, switchToAdvancedEditor, fetchAdvancedSettings };
