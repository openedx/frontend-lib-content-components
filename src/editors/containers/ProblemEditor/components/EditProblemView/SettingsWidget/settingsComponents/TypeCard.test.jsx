import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { formatMessage } from '../../../../../../../testUtils';
import { TypeCard } from './TypeCard';
import { ProblemTypeKeys } from '../../../../../../data/constants/problem';

describe('TypeCard', () => {
  const props = {
    answers: [],
    blockTitle: 'BLocktiTLE',
    correctAnswerCount: 0,
    problemType: ProblemTypeKeys.TEXTINPUT,
    setBlockTitle: jest.fn().mockName('args.setBlockTitle'),
    updateField: jest.fn().mockName('args.updateField'),
    updateAnswer: jest.fn().mockName('args.updateAnswer'),
    // injected
    intl: { formatMessage },
  };

  describe('snapshot', () => {
    test('snapshot: renders type setting card', () => {
      expect(shallow(<TypeCard {...props} />).snapshot).toMatchSnapshot();
    });
  });
});
