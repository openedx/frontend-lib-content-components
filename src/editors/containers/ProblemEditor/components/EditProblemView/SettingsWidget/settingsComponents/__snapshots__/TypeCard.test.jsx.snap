// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeCard snapshot snapshot: renders type setting card 1`] = `
<SettingsOption
  className=""
  none={false}
  summary="Text input"
  title="Type"
>
  <TypeRow
    answers={Array []}
    correctAnswerCount={0}
    key="multiplechoiceresponse"
    label="Single select"
    lastRow={false}
    selected={true}
    typeKey="multiplechoiceresponse"
    updateAnswer={[MockFunction args.updateAnswer]}
    updateField={[MockFunction args.updateField]}
  />
  <TypeRow
    answers={Array []}
    correctAnswerCount={0}
    key="choiceresponse"
    label="Multi-select"
    lastRow={false}
    selected={true}
    typeKey="choiceresponse"
    updateAnswer={[MockFunction args.updateAnswer]}
    updateField={[MockFunction args.updateField]}
  />
  <TypeRow
    answers={Array []}
    correctAnswerCount={0}
    key="optionresponse"
    label="Dropdown"
    lastRow={false}
    selected={true}
    typeKey="optionresponse"
    updateAnswer={[MockFunction args.updateAnswer]}
    updateField={[MockFunction args.updateField]}
  />
  <TypeRow
    answers={Array []}
    correctAnswerCount={0}
    key="numericalresponse"
    label="Numerical input"
    lastRow={false}
    selected={true}
    typeKey="numericalresponse"
    updateAnswer={[MockFunction args.updateAnswer]}
    updateField={[MockFunction args.updateField]}
  />
  <TypeRow
    answers={Array []}
    correctAnswerCount={0}
    key="stringresponse"
    label="Text input"
    lastRow={true}
    selected={false}
    typeKey="stringresponse"
    updateAnswer={[MockFunction args.updateAnswer]}
    updateField={[MockFunction args.updateField]}
  />
</SettingsOption>
`;
